Fixed progress page update on save ccx and grade book crash issue

amir-qayyum-khan opened the pull request:
### Background
fixes https://github.com/mitocw/edx-platform/issues/252 https://openedx.atlassian.net/browse/TNL-4892

### What is done in this PR
**Studio Updates:** None.
**LMS Updates:** 
 - Add signal on save_ccx api, So that other application can be up to date when ccx contents change,

@pdpinch @giocalitri 

### To test:
- Create a coach
- create a student
- Login to edX platform as coach, open coach dashboard. change schedule of subsection aka sequential and press save.
<img width="615" alt="screen shot 2016-07-14 at 7 32 14 pm" src="https://cloud.githubusercontent.com/assets/10431250/16843215/cd27feea-49f9-11e6-9326-56e91ae61a5a.png">
<img width="757" alt="screen shot 2016-07-14 at 7 32 23 pm" src="https://cloud.githubusercontent.com/assets/10431250/16843216/cd2b91fe-49f9-11e6-83bf-ec95ef18c708.png">

- Login as student (CCX enrolled student). Go to dashboard then open ccx course and go to progress and see the due date of that subsection(problem).
<img width="1036" alt="screen shot 2016-07-14 at 7 34 41 pm" src="https://cloud.githubusercontent.com/assets/10431250/16843404/8c3e5a0e-49fa-11e6-8bbc-30b4c59773cd.png">
<img width="971" alt="screen shot 2016-07-14 at 7 34 52 pm" src="https://cloud.githubusercontent.com/assets/10431250/16843408/8cfca5ae-49fa-11e6-8024-c25fad06b69a.png">
<img width="653" alt="screen shot 2016-07-14 at 7 37 25 pm" src="https://cloud.githubusercontent.com/assets/10431250/16843409/8d27ef48-49fa-11e6-85f9-5c672c6f6250.png">

edx-webhook commented:
Thanks for the pull request, @amir-qayyum-khan! I've created [OSPR-1346](https://openedx.atlassian.net/browse/OSPR-1346) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

nedbat commented on lms/djangoapps/ccx/tests/test_views.py:
```diff
@@ -131,6 +134,16 @@ def setup_students_and_grades(context):
 
+def unhide(unit):
+    """
+    Recursively unhide a unit and all of its children in the CCX
```
I know this was a pre-existing function, but why recursion? A simple loop would do! :)

nedbat commented:
@cahrens @doctoryes CCX stuff often goes to platform, but this seems to affect touch UI, I wonder if TNL should take it?

cahrens commented:
I can take it-- I've reviewed #11647.

cahrens commented:
Though note that there are only Python files in this PR... So if @doctoryes WANTS to take the PR, I'm fine with that.

cahrens commented:
And furthermore, @nasthagiri may wish to review since she opened the bug initially.

doctoryes commented:
@cahrens I will review it.

cahrens commented:
Great!

doctoryes commented:
LGTM. One question:
Are there other places that need to send the course_published signal? For example, there's a `create_ccx` view - why does it and potentially other create/update operations not need to send this signal for other parts of the system to consume?

pdpinch commented:
That's a good question. 

@amir-qayyum-khan can you confirm that adding this signal to save_ccx catches all potential changes, i.e., changes to the schedule or the grading policy?

The CCX isn't very interested at the point of create_ccx, but it seems like we should add the signal there as well for completeness. 

I can't think of any other places whether the coach can modify the ccx overrides.

amir-qayyum-khan commented:
yes i verified changes for schedule save but need to double check policy
because there is another api in ccx app to save grading policy.

No need to add signal on ccx create api,  because when ccx create after
that we enroll students and they see all updates in ccx.

On Jul 26, 2016 9:12 PM, "Peter Pinch" <notifications@github.com> wrote:

> That's a good question.
>
> @amir-qayyum-khan <https://github.com/amir-qayyum-khan> can you confirm
> that adding this signal to save_ccx catches all potential changes, i.e.,
> changes to the schedule or the grading policy?
>
> The CCX isn't very interested at the point of create_ccx, but it seems
> like we should add the signal there as well for completeness.
>
> I can't think of any other places whether the coach can modify the ccx
> overrides.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/edx/edx-platform/pull/13019#issuecomment-235319405>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJ8rEhNmJv6ztg5IyHl5vPunEu5O1cCpks5qZjHRgaJpZM4JO0Se>
> .
>

amir-qayyum-khan commented:
@doctoryes made some changes

1) Added signal in all paths where ccx course is change including rest api
2) fixed issue related to gradebook and csv files. Made ccx id unicode to fix grade book crash issue when getting ```course_structure = get_course_blocks(student, course.location)```
3) add test case

cc @pdpinch

amir-qayyum-khan commented:
jenkins run bokchoy

doctoryes commented on lms/djangoapps/ccx/views.py:
```diff
@@ -345,6 +363,14 @@ def set_grading_policy(request, course, ccx=None):
+
     url = reverse(
         'ccx_coach_dashboard',
         kwargs={'course_id': CCXLocator.from_course_locator(course.id, ccx.id)}
```
Why does the signal line above need a `unicode(ccx.id)` but this `course_id` construction does not?

amir-qayyum-khan replied:
it should be, but CCXLocator with non unicode also works, it only give issues when we compare same CCXLocator one with ccx.id unicode and one with not.

let me fix through out views

doctoryes commented:
While this change seems good enough to merge (after tests pass), I'd like to see some follow-on work:
- Check all instances of this construction to determine whether `ccx.id` needs to be wrapped in `unicode()`:
  - `ccx_key = CCXLocator.from_course_locator(course_key, ccx.id)`
  - For example, this location (and others) don't do the wrapping:
    - https://github.com/edx/edx-platform/blob/master/lms/djangoapps/ccx/tasks.py#L36
- Determine whether tests already exist for checking the correct number of signals are emitted when performing the operations for which course_published signals were added.
  - No tests were added which confirm that the added signals are sent at all for some operations and the correct number of signals are sent.

@pdpinch Thoughts?

P.S. Ah - I see @amir-qayyum-khan has already submitted a commit partially addressing my first point (while I was typing this list).

amir-qayyum-khan commented:
jenkins run bokchoy

amir-qayyum-khan commented:
@doctoryes  all test are complete, can u give thumbs up on this?

doctoryes commented:
👍

amir-qayyum-khan commented:
@doctoryes done with verification on stage.

doctoryes commented:
@amir-qayyum-khan Thanks! I'll mark testing as complete on the release page.

amir-qayyum-khan commented:
thank you @doctoryes
